bisq-network/bisq
# Deterministic/Reproducible build of shaded.jar

It is now possible to reproduce, byte for byte, every build of Bisq. More information about reproducible builds in general are available at https://reproducible-builds.org Reproducible builds are a set of software development practices that create a verifiable path from human readable source code to the binary code used by computers

Why we want that?
* To prove that a given binary file is the result of the compilation of a given source hash
* to increase security

## How to test:
checkout this branch, and run 
`mvn clean package`
then check hashes of shaded jar
```
sha256sum gui\target\shaded.jar seednode\target\SeedNode.jar statistics/target/Statistics.jar provider/target/provider.jar
```
will return

```
0ab3c7d95df4db9af20c0c07e468220348a8e13dd0d1f2d99e4915ef0dd8318b *gui\\target\\shaded.jar
ae80419acdf208a2eae4022af710703b4848659eeca9a7d015f968a176ba87c1 seednode\\target\\SeedNode.jar
4dad338913b99bf93214d5bdcd60dc31c7bce8c7feb0b1ac6e82165a9393c6c2 statistics/target/Statistics.jar
df0310e0f2bef2bf36371f4d39865654df2e81d161c3051ff9a59cdef0c57209 *provider/target/provider.jar
```

run again 
`mvn clean package`
and you will get the same hashes again and again....

# Notes
* I use maven-enforcer-plugin  in order to compile Bisq on a given compiler version (java 1.8 and Maven 3.5.0) and to ensure all plugins dependencies have an imposed version
* a 3rd party plugin is required for now strip-jar: it removes un-reproducible bits inside the JAR (timestamp), sets ZIP format time-stamps to 0 and re-orders the files inside the ZIP so they are always in a deterministic order. 
* Apache Maven has recognized the issue, and some work is being done see https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=74682318 but it no ready yet.
